src/Integration.Vsix.UnitTests/Analysis/IssueHandlerTests.cs

```diff
@@ -0,0 +1,271 @@
+        {
+            var mockSnapshotLine = new Mock<ITextSnapshotLine>();
+            mockSnapshotLine.Setup(x => x.GetText()).Returns(textToReturn);
+
+            var mockSnapshot = new Mock<ITextSnapshot>();
+            mockSnapshot.Setup(x => x.Length).Returns(9999);
+            mockSnapshot.Setup(x => x.LineCount).Returns(lineCount);
+            mockSnapshot.Setup(x => x.GetLineFromLineNumber(It.IsAny<int>()))
+                .Returns(mockSnapshotLine.Object);
+
+            return mockSnapshot;
+        }
+
+        private static IAnalysisIssueVisualization CreateFileLevelIssue(string ruleKey)
+        {
+            var issue = new DummyAnalysisIssue
+            {
+                RuleKey = ruleKey,
+                PrimaryLocation = new DummyAnalysisIssueLocation { TextRange = null }
+            };
+
+            var issueVizMock = new Mock<IAnalysisIssueVisualization>();
+            issueVizMock.Setup(x => x.Issue).Returns(issue);
+            issueVizMock.Setup(x => x.Location).Returns(issue.PrimaryLocation);
+            issueVizMock.Setup(x => x.Flows).Returns(Array.Empty<IAnalysisIssueFlowVisualization>());
+
+            return issueVizMock.Object;
+        }
+
+        private static ITextDocument CreateValidTextDocument(string filePath)
+        {
+            var document = new Mock<ITextDocument>();
+            var buffer = new Mock<ITextBuffer>();
+            var snapshot = new Mock<ITextSnapshot>();
+
+            document.Setup(x => x.FilePath).Returns(filePath);
+            document.Setup(x => x.TextBuffer).Returns(buffer.Object);
+            buffer.Setup(x => x.CurrentSnapshot).Returns(snapshot.Object);
+
+            return document.Object;
+        }
+
+        private static IssueHandler CreateTestSubject(IIssuesFilter issuesFilter,
+            PublishSnapshot publishSnapshot)
+            => CreateTestSubject(issuesFilter, publishSnapshot, "any project name", Guid.NewGuid(), "any file.txt");
+
+        private static IssueHandler CreateTestSubject(IIssuesFilter issuesFilter,
+            PublishSnapshot publishSnapshot, string projectName, Guid projectGuid, string filePath)
+        {
+            issuesFilter ??= Mock.Of<IIssuesFilter>();
+            publishSnapshot ??= Mock.Of<PublishSnapshot>();
+
+            var testSubject = new IssueHandler(
+                CreateValidTextDocument(filePath),
+                projectName,
+                projectGuid,                
+                issuesFilter,
+                publishSnapshot,
+                // Override the un-testable "TranslateSpans" behaviour
+                PassthroughSpanTranslator);
+
+            return testSubject;
+        }
+
+        private static IEnumerable<IAnalysisIssueVisualization> PassthroughSpanTranslator(IEnumerable<IAnalysisIssueVisualization> issues, ITextSnapshot activeSnapshot)
+            => issues;
+
+        internal class SnapshotPublisher
```

Comment:
Helper class. Simpler and easier to read than using mocks to check the delegate invocation and parameters.